### DISCUSSION TITLE
Handling migration

### DISCUSSION THREAD

**Discussion by rickmoonex** [thingsdb/ThingsDB#388] [Q&A] [answered]
I was wondering how you guys tackle migrations within ThingsDB. Like creating/updating types, procedures, or enums.

Been struggling a bit with getting a structured way of handling this.

**Comment by joente** [comment, marked as answer]
Let me describe our workflow but keep in mind that this might be improved, and we're open to ideas for tools which might help improve the workflow.

We use a "dump" of all the types, enum's and procedures in *.ti files in a single location (a single file might also work). For changes we create a sub branch, for example `issue001` and when we need changes we apply them both to the corresponding ti file, and just before we merge the branch we create a corresponding `issue001.ti` migration file in a `migration` folder. 

In the `issue001.ti` we define the migration code which can be a new procedure, but also a change to a procedure (`mod_procedure`) or a change to an enumerator or type.

In our IDE's, we usually exclude the `/migration` folder from search&replace as we want to keep those immutable. In our case we also remove old migration files, although this is not necessary.

Another approach might be to work with migration files only, but the downside is that it might be hard to find the current state as this might be spread across  different files and therefore would require to look in ThingsDB for the actual definitions etc.
